src/Symfony/Component/Mime/MimeTypes.php

```diff
@@ -0,0 +1,1899 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mime;
+
+/**
+ * Manages mime types and file extensions.
+ *
+ * For mime type guessing, you can register custom guessers
+ * by calling the registerGuesser() method.
+ * Custom guessers are always called before any default ones:
+ *
+ *     $guesser = new MimeTypes();
+ *     $guesser->registerGuesser(new MyCustomMimeTypeGuesser());
+ *
+ * If you want to change the order of the default guessers, just re-register your
+ * preferred one as a custom one. The last registered guesser is preferred over
+ * previously registered ones.
+ *
+ * Re-registering a built-in guesser also allows you to configure it:
+ *
+ *     $guesser = new MimeTypes();
+ *     $guesser->registerGuesser(new FileinfoMimeTypeGuesser('/path/to/magic/file'));
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+final class MimeTypes implements MimeTypesInterface
+{
+    private $guessers = [];
+    private static $default;
+
+    public function __construct(array $map = [])
```

Comment:
The way this works looks a bit "un-Symfony" to me because of this:

* We register the guessers unconditionally
* We don't allow to change or define their priorities
* We check if a guesser is supported in runtime (with some expensive function calls)

Instead, this would look more Symfony-ish:

* Register guessers as tagged services with priorities
* Run the `isGuesserSupported()` method while registering them, to run that expensive call once instead of everytime a MIME operation is performed, and remove unsupported guessers

Reply:
We run `isGuesserSupported()` as late as possible as one of the guesser is very expensive to run (we changed that a while ago as a perf opt). For such expensive guessers, we have a local cache which means that additionnal checks are really fast.

Reply:
Regarding priorities, we indeed register 2 **default** guessers automatically with a pre-defined order that you don't want to change (performance-wise). You can define your own which is going to take precedence over the other ones. And you should very rarely need to register your own, except for some very specific use case (read: no third-party bundles should do that). So, we will only have exactly 1 custom guesser (so priority is not needed here).